### PR TITLE
RavenDB-16595 - Unable to expand database group

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-16595/Unable-to-expand-database-group

### Additional description

Fix - not abling to expand database group (The restored db stays 'promotable' after adding another node to it).
The reason for this is that the ClusterObserver waits for the documents Etag (of the existing database) to be the same as the database Etag in order to change the state of the new node (the one that you're about to add to this DB) from `promotable` to `member`, but there are cases when the documents Etag can be smaller than the database Etag (for example: when the DB had tombstones and it cleaned them).
The node that you want to add to the DB doesn't become 'member' (stays 'promotable') because the sender (the 'member' which sent the documents to the 'promotable') doesn't get its last etag
(in method ReplicationDocumentSender.lastEtagFromDestinationChangeVector - var lastEtagFromDestinationChangeVector), this happens because the change vector contains etag with the old database id (the one that backed up to the snapshot) and doesn't contains etag with the new (restored) database id, which is different.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: test/Tests.Infrastructure/RavenTestBase.Backup.cs [hunk 1]
```diff
@@ -227,7 +228,11 @@ public async Task RunBackupInClusterAsync(DocumentStore store, long taskId, bool
 
             public IDisposable RestoreDatabase(IDocumentStore store, RestoreBackupConfiguration config, TimeSpan? timeout = null)
             {
-                var restoreOperation = new RestoreBackupOperation(config);
+                RestoreBackupOperation restoreOperation;
+                if (config.NodeTag != null)
+                    restoreOperation = new RestoreBackupOperation(config, config.NodeTag);
+                else
+                    restoreOperation = new RestoreBackupOperation(config);
```
Comment (karmeli87) [inline test/Tests.Infrastructure/RavenTestBase.Backup.cs]: it is okay to have this, but why is that required for the test?

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 2]
```diff
@@ -1349,22 +1349,29 @@ private bool TryGetMentorNode(string dbName, DatabaseTopology topology, ClusterT
...
-            if (mentorCurrDbStats.LastSentEtag.TryGetValue(promotable, out var lastSentEtag) == false)
+            var mentorLastDbEtag = mentorPrevDbStats.LastEtag;
+            if (mentorCurrDbStats.LastSentEtag.TryGetValue(promotable, out var mentorLastDocsEtag) == false)
             {
-                LogMessage($"Can't find last sent etag of mentor {mentorNode} for {promotable}", database: dbName);
+                LogMessage($"Can't find last sent documents etag of mentor {mentorNode} for {promotable}", database: dbName);
+                return (false, null);
+            }
+
+            var promotableLastDbEtag = mentorPrevDbStats.LastEtag;
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: how is that `promotableLastDbEtag` this is identical to `mentorLastDbEtag`

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 3]
```diff
@@ -1349,22 +1349,29 @@ private bool TryGetMentorNode(string dbName, DatabaseTopology topology, ClusterT
...
+            if (mentorCurrDbStats.LastSentEtag.TryGetValue(promotable, out var promotableLastDocsEtag) == false)
+            {
+                LogMessage($"Can't find last sent documents etag of promotable {promotable}", database: dbName);
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: What do I miss? this is the same as the above, no?

File: src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs [hunk 4]
```diff
@@ -40,6 +40,8 @@ public class RestoreBackupConfiguration : RestoreBackupConfigurationBase
     {
         public string BackupLocation { get; set; }
 
+        public string NodeTag { get; set; }
```
Comment (grisha-kotler) [inline src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs]: Why do you need the `NodeTag` here? (this isn't used in the restore process)
You can add another parameter to `RestoreDatabase`.

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs (hunk 3 above)
Comment (shaharhikri) [reply src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: You're right, I changed it to 'promotablePrevDbStats' and 'promotablePrevDbStats'

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: You're right, I changed it to 'promotablePrevDbStats' and 'promotablePrevDbStats'

File: src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs (hunk 4 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs]: I want to restore it to a specific node.

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs (hunk 2 above)
Comment (shaharhikri) [reply src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: I want to restore it to a specific node.

File: test/Tests.Infrastructure/RavenTestBase.Backup.cs (hunk 1 above)
Comment (shaharhikri) [reply test/Tests.Infrastructure/RavenTestBase.Backup.cs]: I want to restore it to a specific node.

File: src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs]: It's already handled at the operation level

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 5]
```diff
@@ -1349,22 +1349,29 @@ private bool TryGetMentorNode(string dbName, DatabaseTopology topology, ClusterT
...
+            var mentorLastDbEtag = mentorPrevDbStats.LastEtag;
+            if (mentorCurrDbStats.LastSentEtag.TryGetValue(promotable, out var mentorLastDocsEtag) == false)
             {
-                LogMessage($"Can't find last sent etag of mentor {mentorNode} for {promotable}", database: dbName);
+                LogMessage($"Can't find last sent documents etag of mentor {mentorNode} for {promotable}", database: dbName);
+                return (false, null);
+            }
+
+            var promotableLastDbEtag = promotablePrevDbStats.LastEtag;
+            if (promotablePrevDbStats.LastSentEtag.TryGetValue(promotable, out var promotableLastDocsEtag) == false)
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: why do we care about `promotablePrevDbStats.LastSentEtag` ? 
This is the last etag that was _sent_ by the promotable

File: src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs (hunk 4 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Operations/Backups/RestoreBackupConfiguration.cs]: Removed this property.

**Review by grisha-kotler — APPROVED** [review]

**Comment by arekpalinski** [comment]
test this please

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: I removed this part, it's not relevant

**Review by karmeli87 — APPROVED** [review]
